### PR TITLE
General sharded persister

### PR DESCRIPTION
**Pull request by ssd04**

### CODE REVIEW

**Review by BeniaminDrasovean — DISMISSED** [review]

File: sharded/shardIDProvider_test.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func TestGetShardIDs(t *testing.T) {
 func TestComputeId(t *testing.T) {
```
Comment (iulianpascalau) [inline sharded/shardIDProvider_test.go:59]: can we have a similar test for this but with 5 shards?

File: sharded/shardedDB.go [hunk 2]
```diff
@@ -40,8 +43,19 @@ func NewShardedPersister(path string, batchDelaySeconds int, maxBatchSize int, m
 	}, nil
 }
 
+func createDB(dbType storageUnit.DBType, path string, batchDelaySeconds int, maxBatchSize int, maxOpenFilesPerShard int) (types.Persister, error) {
```
Comment (iulianpascalau) [inline sharded/shardedDB.go]: by having this function, there is no other way to reuse this component with other types other than serialDB or DB. If we will implement a new persister type, we will need to change this function also. This breaks the OCP principle.
Solution: on the NewShardedPersister you should provide a factory that can create a new persister along with the ShardIDProvider. Something like:
```GO
NewShardedPersister(persisterFactory PersisterFactory, idProvider types.ShardIDProvider)
// .... 

type PersisterFactory interface{
    Create(newPath) (types.Persister, error)
    GetBasePath() string
}
```

we can then implement this persister factory that will contain the switch inside it. The rest needed parameters can be provided on the persister factory's constructor.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: sharded/shardIDProvider.go [hunk 3]
```diff
@@ -5,6 +5,7 @@ import (
 	"math"
 )
 
+const minNumOfShards = 1
```
Comment (iulianpascalau) [inline sharded/shardIDProvider.go]: does it make sense to have 1 shard in shard ID provider? :D 
I suggest to change this to 2

Comment (ssd04) [reply sharded/shardIDProvider.go]: changed to 2 :+1:

File: sharded/shardedDB_test.go [hunk 4]
```diff
@@ -16,15 +16,25 @@ func TestNewShardedPersister(t *testing.T) {
 	t.Run("nil persister creator", func(t *testing.T) {
```
Comment (iulianpascalau) [inline sharded/shardedDB_test.go]: empty path should error

**Review by iulianpascalau — DISMISSED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

File: storageUnit/storageunit.go [hunk 5]
```diff
@@ -34,7 +34,7 @@ type DBType string
 // HasherType represents the type of the supported hash functions
 type HasherType string
 
-// LRUCache is currently the only supported Cache type
+// Cache types that are currently supported
```
Comment (iulianpascalau) [inline storageUnit/storageunit.go:37]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**